### ISSUE TITLE
Can't upgrade from 0.11.14 to 0.13.0 with version 3 statefile containing remote state data resource

### ISSUE THREAD

**Issue by alisdair** [hashicorp/terraform#25863]
### Terraform Version

Using the branch from #25861

```
Terraform v0.13.1-dev
```

### Terraform Configuration Files
```terraform
data "terraform_remote_state" "test" {
  backend = "remote"

  config = {
    organization = "my-tfc-org"
    workspaces = {
      name = "some-tfc-workspace"
    }
  }
}
```

### Debug Output
[Log gist](https://gist.github.com/alisdair/b74fef49082fb572760c50ea7e752663)

### Expected Behavior

An empty plan.

### Actual Behavior

```shellsession
$ terraform plan
Refreshing Terraform state in-memory prior to plan...
The refreshed state will be used to calculate this plan, but will not be
persisted to local or remote state storage.


Error: cannot decode dynamic from flatmap

$
```

### Steps to Reproduce

1. Create a config with a Terraform Cloud remote state data source block pointing to a workspace
1. Run `terraform init` and `terraform apply` with Terraform 0.11.14 to create a state file
1. Run `terraform plan` with Terraform 0.13.1-dev (after #25861 is merged, or using that branch, or it'll panic)

### Additional Context

This works with Terraform 0.12.28. The failure is in attempting to decode the dynamic `config` block from the flatmap in the state file. From looking at the v0.12 branch, I can't see how this used to work, nor any reason why it should only start failing with 0.13.

### References

Found when investigating the fix for #25803.

**Comment by apparentlymart** [comment]
In your "steps to reproduce" it seems like you went directly from 0.11.14 to 0.13.1 without applying with a 0.12.x release first. Does this still fail if you upgrade only one major version at a time?

Our typical policy is that we don't support skipping major releases, because that creates the risk of missing out on upgrade steps that were present in an earlier release but no longer present in the current release. So if this is "fixed" by running `terraform apply` with 0.12.x first then I'd be inclined to consider it a "won't fix".

**Comment by alisdair** [comment]
That's correct—running a `terraform apply` with 0.12 upgrades the state file and 0.13 is then able to use it. The policy makes sense to me, closing!

**Comment by unacceptable** [comment]
@alisdair Thanks for the update... I will give that a go in the morning.
